source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -0,0 +1,286 @@
+  std::string uri_;
+  // The pending remote request so it can be canceled.
+  Http::AsyncClient::Request* request_{};
+};
+
+void AuthenticatorImpl::sanitizePayloadHeaders(Http::HeaderMap& headers) {
+  for (const auto& rule : store_.config().rules()) {
+    if (!rule.forward_payload_header().empty()) {
+      headers.remove(Http::LowerCaseString(rule.forward_payload_header()));
+    }
+  }
+}
+
+void AuthenticatorImpl::verify(Http::HeaderMap& headers, Authenticator::Callbacks* callback) {
+  headers_ = &headers;
+  callback_ = callback;
+
+  ENVOY_LOG(debug, "Jwt authentication starts");
+  auto tokens = store_.getExtractor().extract(headers);
+  if (tokens.size() == 0) {
+    if (okToBypass()) {
+      doneWithStatus(Status::Ok);
+    } else {
+      doneWithStatus(Status::JwtMissed);
+    }
+    return;
+  }
+
+  // TODO, add supports for multiple tokens.
+  // Only process the first token for now.
+  token_.swap(tokens[0]);
+
+  Status status = jwt_.parseFromString(token_->token());
+  if (status != Status::Ok) {
+    doneWithStatus(status);
+    return;
+  }
+
+  // Check if token is extracted from the location specified by the issuer.
+  if (!token_->isIssuerSpecified(jwt_.iss_)) {
+    ENVOY_LOG(debug, "Jwt for issuer {} is not extracted from the specified locations", jwt_.iss_);
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check "exp" claim.
+  const auto unix_timestamp = std::chrono::duration_cast<std::chrono::seconds>(
+                                  std::chrono::system_clock::now().time_since_epoch())
+                                  .count();
+  if (jwt_.exp_ < unix_timestamp) {
+    doneWithStatus(Status::JwtExpired);
+    return;
+  }
+
+  // Check the issuer is configured or not.
+  jwks_data_ = store_.getJwksCache().findByIssuer(jwt_.iss_);
+  if (jwks_data_ == nullptr) {
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check if audience is allowed
+  if (!jwks_data_->areAudiencesAllowed(jwt_.audiences_)) {
+    doneWithStatus(Status::JwtAudienceNotAllowed);
+    return;
+  }
+
+  if (jwks_data_->getJwksObj() != nullptr && !jwks_data_->isExpired()) {
+    verifyKey();
+    return;
+  }
+
+  fetchRemoteJwks();
+}
+
+void AuthenticatorImpl::fetchRemoteJwks() {
+  uri_ = jwks_data_->getJwtRule().remote_jwks().http_uri().uri();
+  std::string host, path;
+  extractUriHostPath(uri_, &host, &path);
+
+  Http::MessagePtr message(new Http::RequestMessageImpl());
+  message->headers().insertMethod().value().setReference(Http::Headers::get().MethodValues.Get);
+  message->headers().insertPath().value(path);
+  message->headers().insertHost().value(host);
+
+  const auto& cluster = jwks_data_->getJwtRule().remote_jwks().http_uri().cluster();
+  if (cm_.get(cluster) == nullptr) {
+    doneWithStatus(Status::JwksFetchFail);
+    return;
+  }
+
+  ENVOY_LOG(debug, "fetch pubkey from [uri = {}]: start", uri_);
+  request_ = cm_.httpAsyncClientForCluster(cluster).send(
```

Comment:
Is the assumption here that the cluster will be configured for TLS? I'd be interesting in seeing some way to validate or warn if this isn't the case, since it would be a pretty bad misconfiguration to fetch certs or public keys without TLS.

Reply:
Normally, public key servers only support TLS.  Users have to config cluster with TLS.

Reply:
I'm OK with saying this, but I think it would make sense to document with a strong warning here. It would be pretty scary to get this one wrong.